ml-agents/mlagents/trainers/trajectory.py

```diff
@@ -0,0 +1,128 @@
+from typing import List, NamedTuple
+import numpy as np
+
+from mlagents.trainers.buffer import AgentBuffer
+
+
+class AgentExperience(NamedTuple):
+    obs: List[np.ndarray]
+    reward: float
+    done: bool
+    action: np.ndarray
+    action_probs: np.ndarray
+    action_pre: np.ndarray  # TODO: Remove this
+    action_mask: np.ndarray
+    prev_action: np.ndarray
+    max_step: bool
+    memory: np.ndarray
+
+
+class SplitObservations(NamedTuple):
+    vector_observations: np.ndarray
+    visual_observations: List[np.ndarray]
+
+    @staticmethod
+    def from_observations(obs: List[np.ndarray]) -> "SplitObservations":
+        """
+        Divides a List of numpy arrays into a SplitObservations NamedTuple.
+        This allows you to access the vector and visual observations directly,
+        without enumerating the list over and over.
+        :param obs: List of numpy arrays (observation)
+        :returns: A SplitObservations object.
+        """
+        vis_obs_indices = []
+        vec_obs_indices = []
+        for index, observation in enumerate(obs):
+            if len(observation.shape) == 1:
+                vec_obs_indices.append(index)
+            if len(observation.shape) == 3:
+                vis_obs_indices.append(index)
+        vec_obs = (
+            np.concatenate([obs[i] for i in vec_obs_indices], axis=0)
+            if len(vec_obs_indices) > 0
+            else np.array([], dtype=np.float32)
+        )
+        vis_obs = [obs[i] for i in vis_obs_indices]
+        return SplitObservations(
+            vector_observations=vec_obs, visual_observations=vis_obs
+        )
+
+
+class Trajectory(NamedTuple):
+    steps: List[AgentExperience]
+    next_obs: List[
+        np.ndarray
+    ]  # Observation following the trajectory, for bootstrapping
+    agent_id: str
+
+    def to_agentbuffer(self) -> AgentBuffer:
+        """
+        Converts a Trajectory to an AgentBuffer
+        :param trajectory: A Trajectory
+        :returns: AgentBuffer. Note that the length of the AgentBuffer will be one
+        less than the trajectory, as the next observation need to be populated from the last
+        step of the trajectory.
+        """
+        agent_buffer_trajectory = AgentBuffer()
+        for step, exp in enumerate(self.steps):
+            vec_vis_obs = SplitObservations.from_observations(exp.obs)
```

Comment:
Note that you're calling SplitObservations.from_observations twice on each element (except maybe the start and end). Might be worth changing the loop to something like
```
vec_vis_obs = SplitObservations.from_observations(self.steps[0])
for step, exp in enumerate(self.steps):
  # get next_vec_vis_obs same as now
  # rest of loop
  vec_vis_obs = next_vec_vis_obs
```
but then again, from_observations looks pretty fast unless you're concatenating large vector observations.

Reply:
Yep, fixed